File: frontend/src/api/cad.ts
import { apiClient, ApiResponse } from './client';
import type { FileInfo } from './file';

export type CADUnit = 'mm' | 'inch';

export interface CADPoint {
  x: number;
  y: number;
}

export interface CADLayer {
  name: string;
  color?: number;
  line_type?: string;
  visible: boolean;
}

export interface CADLine {
  id: string;
  type: 'line';
  layer: string;
  start: CADPoint;
  end: CADPoint;
}

export interface CADCircle {
  id: string;
  type: 'circle';
  layer: string;
  center: CADPoint;
  radius: number;
}

export interface CADArc {
  id: string;
  type: 'arc';
  layer: string;
  center: CADPoint;
  radius: number;
  start_angle: number;
  end_angle: number;
}

export interface CADPolyline {
  id: string;
  type: 'polyline';
  layer: string;
  points: CADPoint[];
  closed: boolean;
}

export interface CADSlot {
  id: string;
  type: 'slot';
  layer: string;
  center: CADPoint;
  width: number;
  length: number;
  rotation: number;
}

export interface CADNote {
  id: string;
  type: 'note';
  layer: string;
  position: CADPoint;
  text: string;
  height: number;
}

export type CADEntity = CADLine | CADCircle | CADArc | CADPolyline | CADSlot | CADNote;

export interface CADDimension {
  id: string;
  type: 'linear' | 'diameter' | 'radius' | 'note';
  layer: string;
  start?: CADPoint;
  end?: CADPoint;
  position: CADPoint;
  text?: string;
}

export interface MechanicalDesignBrief {
  part_type?: string;
  units: CADUnit;
  features: Record<string, any>[];
  constraints: string[];
  unknowns: string[];
  manufacturing_notes: string[];
  source_references: Record<string, any>[];
}

export interface MechanicalCADDocument {
  id: string;
  session_id?: string;
  user_id?: string;
  title: string;
  units: CADUnit;
  layers: CADLayer[];
  entities: CADEntity[];
  dimensions: CADDimension[];
  constraints: Record<string, any>[];
  brief?: MechanicalDesignBrief;
  version: number;
  created_at: string;
  updated_at: string;
}

export interface CADOperation {
  operation: string;
  params: Record<string, any>;
}

export interface CADDocumentResponse {
  document: MechanicalCADDocument;
}

export interface ApplyCADOperationResponse {
  document: MechanicalCADDocument;
  added_entity_ids: string[];
  message: string;
}

export interface CADPlanStep {
  id: string;
  title: string;
  description: string;
  operation: CADOperation;
}

export interface CADPlanFromPromptResponse {
  title: string;
  message: string;
  brief: MechanicalDesignBrief;
  steps: CADPlanStep[];
}

export async function createCADDocumentFromPrompt(prompt: string, attachments: FileInfo[] = []): Promise<ApplyCADOperationResponse> {
  const response = await apiClient.post<ApiResponse<ApplyCADOperationResponse>>('/cad/documents/from-prompt', {
    prompt,
    units: 'mm',
    attachments: attachments.map(file => ({
      file_id: file.file_id,
      filename: file.filename,
      content_type: file.content_type,
      size: file.size,
      upload_date: file.upload_date,
    })),
  });
  return response.data.data;
}

export async function createCADDocument(title = 'Mechanical drawing'): Promise<CADDocumentResponse> {
  const response = await apiClient.post<ApiResponse<CADDocumentResponse>>('/cad/documents', {
    title,
    units: 'mm',
  });
  return response.data.data;
}

export async function createCADPlanFromPrompt(prompt: string, attachments: FileInfo[] = []): Promise<CADPlanFromPromptResponse> {
  const response = await apiClient.post<ApiResponse<CADPlanFromPromptResponse>>('/cad/plans/from-prompt', {
    prompt,
    units: 'mm',
    attachments: attachments.map(file => ({
      file_id: file.file_id,
      filename: file.filename,
      content_type: file.content_type,
      size: file.size,
      upload_date: file.upload_date,
    })),
  });
  return response.data.data;
}

export async function applyCADOperation(documentId: string, operation: CADOperation): Promise<ApplyCADOperationResponse> {
  const response = await apiClient.post<ApiResponse<ApplyCADOperationResponse>>(`/cad/documents/${documentId}/operations`, {
    operation,
  });
  return response.data.data;
}

export async function applyCADOperations(documentId: string, operations: CADOperation[]): Promise<ApplyCADOperationResponse> {
  const response = await apiClient.post<ApiResponse<ApplyCADOperationResponse>>(`/cad/documents/${documentId}/operations/batch`, {
    operations,
  });
  return response.data.data;
}

export async function downloadCADDocumentDxf(documentId: string): Promise<Blob> {
  const response = await apiClient.get(`/cad/documents/${documentId}/export/dxf`, {
    responseType: 'blob',
  });
  return response.data;
}

File: backend/app/domain/models/cad.py
from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Literal, Optional, Union
import uuid

from pydantic import BaseModel, Field


class CADUnit(str, Enum):
    MM = "mm"
    INCH = "inch"


class CADLayer(BaseModel):
    name: str
    color: Optional[int] = None
    line_type: Optional[str] = None
    visible: bool = True


class CADPoint(BaseModel):
    x: float
    y: float


class CADEntityBase(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    layer: str = "M-OBJECT"


class CADLine(CADEntityBase):
    type: Literal["line"] = "line"
    start: CADPoint
    end: CADPoint


class CADCircle(CADEntityBase):
    type: Literal["circle"] = "circle"
    center: CADPoint
    radius: float


class CADArc(CADEntityBase):
    type: Literal["arc"] = "arc"
    center: CADPoint
    radius: float
    start_angle: float
    end_angle: float


class CADPolyline(CADEntityBase):
    type: Literal["polyline"] = "polyline"
    points: List[CADPoint]
    closed: bool = False


class CADSlot(CADEntityBase):
    type: Literal["slot"] = "slot"
    center: CADPoint
    width: float
    length: float
    rotation: float = 0


class CADNote(CADEntityBase):
    type: Literal["note"] = "note"
    position: CADPoint
    text: str
    height: float = 3.5


CADEntity = Union[CADLine, CADCircle, CADArc, CADPolyline, CADSlot, CADNote]


class CADDimension(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    type: Literal["linear", "diameter", "radius", "note"] = "linear"
    layer: str = "M-DIM"
    start: Optional[CADPoint] = None
    end: Optional[CADPoint] = None
    position: CADPoint
    text: Optional[str] = None


class CADConstraint(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    type: str
    target_ids: List[str] = Field(default_factory=list)
    params: Dict[str, Any] = Field(default_factory=dict)


class MechanicalDesignBrief(BaseModel):
    part_type: Optional[str] = None
    units: CADUnit = CADUnit.MM
    features: List[Dict[str, Any]] = Field(default_factory=list)
    constraints: List[str] = Field(default_factory=list)
    unknowns: List[str] = Field(default_factory=list)
    manufacturing_notes: List[str] = Field(default_factory=list)
    source_references: List[Dict[str, Any]] = Field(default_factory=list)


class MechanicalCADDocument(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    session_id: Optional[str] = None
    user_id: Optional[str] = None
    title: str = "Untitled mechanical drawing"
    units: CADUnit = CADUnit.MM
    layers: List[CADLayer] = Field(default_factory=list)
    entities: List[CADEntity] = Field(default_factory=list)
    dimensions: List[CADDimension] = Field(default_factory=list)
    constraints: List[CADConstraint] = Field(default_factory=list)
    brief: Optional[MechanicalDesignBrief] = None
    version: int = 1
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class CADOperation(BaseModel):
    operation: str
    params: Dict[str, Any] = Field(default_factory=dict)


class CADOperationResult(BaseModel):
    success: bool
    document: MechanicalCADDocument
    added_entity_ids: List[str] = Field(default_factory=list)
    message: str = ""
